app/src/main/java/org/example/BoardPosition.java

```diff
@@ -0,0 +1,119 @@
+package org.example;
+
+public class BoardPosition {
+    public final int i;
+    public final int j;
+
+    private BoardPosition(int i, int j) {
+        this.i = i;
+        this.j = j;
+    }
+
+    // Singleton class to represent an invalid position. Useful for chaining moves:
+    public static class InvalidBoardPosition extends BoardPosition {
```

Comment:
This is too complicated for what you're trying to do.

The whole point of BoardPosition is to represent only valid positions to avoid redundant checks in the codebase.

You can create moving functions, such as "left", "right", "ahead", "behind" in a way that they don't return a BoardPosition if the move crosses the board boundaries.

One way to do that is to return `Optional<BoardPosition>`. If the optional is empty, it means the position is invalid.

Reply:
Implemented `Optional` across the board.